org.eclipse.xtext.util/src/org/eclipse/xtext/util/UriExtensions.java

```diff
@@ -8,35 +8,103 @@
  */
 package org.eclipse.xtext.util;
 
-import java.net.URISyntaxException;
+import java.util.Arrays;
 
 import org.eclipse.emf.common.util.URI;
 
+import com.google.inject.Singleton;
+
 /**
  * @author Christian Dietrich - Initial contribution and API
  * @since 2.14
  */
+@Singleton
 public class UriExtensions {
 	/**
-	 * Returns an URI with empty authority, if authority is absent (null) and has file scheme.
+	 * Converts the given absolute, encoded stringUri to an EMF URI. If the URI
+	 * is a {@link URI#isFile() file URI}, it is guaranteed that its
+	 * {@link URI#authority() authority} is not null.
 	 */
 	public URI toUri(String stringUri) {
-		try {
-			java.net.URI netUri = new java.net.URI(stringUri);
-			return toEmfUri(netUri);
-		} catch (URISyntaxException e) {
-			throw new RuntimeException(e);
+		URI result = URI.createURI(stringUri);
```

Comment:
Turns out this approach also gives a nice performance boost for URIs that do in fact contain escape sequences.

Simple performance tests show 2 secs -> 0.4 secs = ~80% improvement

Reply:
This is indeed nice.